pkg/identity/gitlabcom/principal.go

```diff
@@ -0,0 +1,205 @@
+	eventName string
+
+	// Pipeline ID
+	pipelineID string
+
+	// Repository building built
+	repository string
+
+	// ID to the source repo
+	repositoryID string
+
+	// Owner of the source repo (mutable)
+	repositoryOwner string
+
+	// ID of the source repo
+	repositoryOwnerID string
+
+	// job ID
+	jobID string
+
+	// Git ref being built
+	ref string
+
+	// Commit SHA being built
+	sha string
+
+	// ID of the runner
+	runnerID int64
+
+	// The type of runner used by the job. May be one of gitlab-hosted or self-hosted.
+	runnerEnvironment string
+}
+
+func JobPrincipalFromIDToken(_ context.Context, token *oidc.IDToken) (identity.Principal, error) {
+	var claims struct {
+		ProjectPath       string `json:"project_path"`
+		ProjectID         string `json:"project_id"`
+		PipelineSource    string `json:"pipeline_source"`
+		PipelineID        string `json:"pipeline_id"`
+		NamespacePath     string `json:"namespace_path"`
+		NamespaceID       string `json:"namespace_id"`
+		JobID             string `json:"job_id"`
+		Ref               string `json:"ref"`
+		RefType           string `json:"ref_type"`
+		Sha               string `json:"sha"`
+		RunnerEnvironment string `json:"runner_environment"`
+		RunnerID          int64  `json:"runner_id"`
+	}
+
+	if err := token.Claims(&claims); err != nil {
+		return nil, err
+	}
+
+	if claims.ProjectPath == "" {
+		return nil, errors.New("missing project_path claim in ID token")
+	}
+
+	if claims.PipelineSource == "" {
+		return nil, errors.New("missing pipeline_source claim in ID token")
+	}
+
+	if claims.PipelineID == "" {
+		return nil, errors.New("missing pipeline_id claim in ID token")
+	}
+
+	if claims.JobID == "" {
+		return nil, errors.New("missing job_id claim in ID token")
+	}
+
+	if claims.Ref == "" {
+		return nil, errors.New("missing ref claim in ID token")
+	}
+
+	if claims.RefType == "" {
+		return nil, errors.New("missing ref_type claim in ID token")
+	}
+
+	if claims.NamespacePath == "" {
+		return nil, errors.New("missing namespace_path claim in ID token")
+	}
+
+	if claims.NamespaceID == "" {
+		return nil, errors.New("missing namespace_id claim in ID token")
+	}
+
+	if claims.ProjectID == "" {
+		return nil, errors.New("missing project_id claim in ID token")
+	}
+
+	if claims.Sha == "" {
+		return nil, errors.New("missing sha claim in ID token")
+	}
+
+	if claims.RunnerEnvironment == "" {
+		return nil, errors.New("missing runner_environment claim in ID token")
+	}
+
+	if claims.RunnerID == 0 {
+		return nil, errors.New("missing runner_id claim in ID token")
+	}
+
+	var ref string
+	switch claims.RefType {
+	case "branch":
+		ref = "refs/heads/" + claims.Ref
+	case "tag":
+		ref = "refs/tags/" + claims.Ref
+	default:
+		return nil, fmt.Errorf("unexpected ref_type: %s", claims.RefType)
+	}
+
+	return &jobPrincipal{
+		subject:           token.Subject,
+		issuer:            token.Issuer,
+		url:               `https://gitlab.com/`,
```

Comment:
The comment for this field says:

```
	// The URL of the pipeline, the container of many builds. This will be
	// set as a human-friendly SubjectAlternativeName URI in the certificate.
	url string
```

To make that comment accurate, this should be

```suggestion
		url:               fmt.Sprintf("https://gitlab.com/%s/-/pipelines/%s", claims.ProjectPath, claims.ProjectID)
```

We could also add another claim for this in the JWT if need be.

Reply:
URL is just the base URL, Embed takes the URL and appends that.

We should update the comment to note that (a previous iteration of Fulcio had the URL as the SAN, some refactoring changed that). Thanks for noticing.